### PR TITLE
object: check obc provisioner for bucket notification

### PR DESCRIPTION
**Pull request by thotz**
**Description of your changes:**
obc-label-controller needs to check only the obc provisioned by the Rook operator.
Check other OBCs result in errors from the controller and keep on reconcilling

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure)).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Review by galexrt — APPROVED** [review]
LGTM

File: pkg/operator/ceph/object/notification/obc_label_controller_test.go [hunk 1]
```diff
@@ -30,11 +30,16 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 )
 
-func createOBResources(name string) (*bktv1alpha1.ObjectBucketClaim, *bktv1alpha1.ObjectBucket) {
+const (
+	testProvisionerOtherCLuster = "other-provisioner.io-bucket"
```
Comment (microyahoo) [inline pkg/operator/ceph/object/notification/obc_label_controller_test.go:34]: CLuster -> Cluster

File: pkg/operator/ceph/object/notification/controller_test.go [hunk 2]
```diff
@@ -57,6 +57,8 @@ var (
 	multipleCreateBucketName = "multi-create"
 	multipleDeleteBucketName = "multi-delete"
 	multipleBothBucketName   = "multi-both"
+	otherClusterBucketName   = "other-cluster"
+	otherNamespaceBucketName = "other-namespace"
```
Comment (subhamkrai) [inline pkg/operator/ceph/object/notification/controller_test.go:61]:
```suggestion
	otherNamespaceBucketName = "other-namespace"
```
not required?

File: pkg/operator/ceph/object/notification/obc_label_controller.go [hunk 3]
```diff
@@ -119,11 +121,17 @@ func (r *ReconcileOBCLabels) reconcile(request reconcile.Request) (reconcile.Res
 	if err := r.client.Get(r.opManagerContext, bucketName, &ob); err != nil {
 		return reconcile.Result{}, errors.Wrapf(err, "failed to retrieve ObjectBucket %q", bucketName)
 	}
+
+	// validate if the bucket is provisioned by the ceph provisioner
+	if !strings.Contains(ob.Labels[bucketProvisionerLabelKey], bucketProvisionerLabelValue) {
```
Comment (travisn) [inline pkg/operator/ceph/object/notification/obc_label_controller.go:126]: If we are just now adding these labels to OBCs, what happens in upgraded clusters where OBCs were previously created without the labels? Does that mean notifications can't be created for OBs created before the upgrade?

**Review by travisn — CHANGES_REQUESTED** [review]

**Comment by yuvalif** [comment]
@thotz maybe a simpler way is to use the error returned from the name parsing function as an indication that the bucket was provisioned by an external provisioner, and let the bucket notification controller ignore it without an error?

File: pkg/operator/ceph/object/notification/obc_label_controller.go (hunk 3 above)
Comment (thotz) [reply pkg/operator/ceph/object/notification/obc_label_controller.go:126]: These labels were present from day one. There are not from Rook. It is from libbucket provisioner

**Review by yuvalif — APPROVED** [review]

File: pkg/operator/ceph/object/notification/obc_label_controller.go (hunk 3 above)
Comment (thotz) [reply pkg/operator/ceph/object/notification/obc_label_controller.go]: label key is not exported by [libbucketprovisioner](https://github.com/kube-object-storage/lib-bucket-provisioner/blob/master/pkg/provisioner/resourcehandlers.go#L51)  and value is populated by Rook, so used same function to compare the values

**Comment by travisn** [comment]
@thotz The object integration tests are failing with this error:
```
=== RUN   TestCephObjectSuite/TestWithTLS/perform_s3_operations_and_check_for_notifications/check_for_put_bucket_notification
2023-03-27 12:10:33.431665 D | exec: Running command: kubectl logs --selector=app=sample-http-server --tail 5
    ceph_bucket_notification_test.go:142: 
        	Error Trace:	/home/runner/work/rook/rook/tests/integration/ceph_bucket_notification_test.go:142
        	Error:      	Should be true
        	Test:       	TestCephObjectSuite/TestWithTLS/perform_s3_operations_and_check_for_notifications/check_for_put_bucket_notification
```

**Comment by BlaineEXE** [comment]
Thanks for the good test updates in the PR. I think it looks good other than the small integration test bug Travis pointed out

**Comment by thotz** [comment]
> @thotz The object integration tests are failing with this error:
> 
> ```
> === RUN   TestCephObjectSuite/TestWithTLS/perform_s3_operations_and_check_for_notifications/check_for_put_bucket_notification
> 2023-03-27 12:10:33.431665 D | exec: Running command: kubectl logs --selector=app=sample-http-server --tail 5
>     ceph_bucket_notification_test.go:142: 
>         	Error Trace:	/home/runner/work/rook/rook/tests/integration/ceph_bucket_notification_test.go:142
>         	Error:      	Should be true
>         	Test:       	TestCephObjectSuite/TestWithTLS/perform_s3_operations_and_check_for_notifications/check_for_put_bucket_notification
> ```

I thought the provisioner's name was the combination of Rook Operator Namespace and "ceph.rook.io/bucket". But the current code uses the ceph cluster namespace. I am confused about how to get that here. In the Bucketprovisioner the are finding the ceph cluster details using the following,
```golang
cephCluster := &cephv1.CephCluster{}
	err := r.client.Get(r.opManagerContext, request.NamespacedName, cephCluster)
	if err != nil {
		if kerrors.IsNotFound(err) {
			logger.Infof("no ceph cluster found in %+v. not deploying the bucket provisioner", request.NamespacedName)
			return reconcile.Result{}, nil
		}
		// Error reading the object - requeue the request.
		return opcontroller.ImmediateRetryResult, errors.Wrap(err, "failed to get the ceph cluster")
	}

	if !cephCluster.DeletionTimestamp.IsZero() {
		logger.Debug("ceph cluster is being deleted, no need to reconcile the bucket provisioner")
		return reconcile.Result{}, nil
	}

	if !cephCluster.Spec.External.Enable && cephCluster.Spec.CleanupPolicy.HasDataDirCleanPolicy() {
		logger.Debug("ceph cluster has cleanup policy, the cluster will soon go away, no need to reconcile the bucket provisioner")
		return reconcile.Result{}, nil
	}
```
The `request.NamespacedName` should be the name space of OBC right and AFAIK obc can exist  any namespace than the ceph objectstore

**Comment by travisn** [comment]
The namespace of the cephcluster is the same as the namespace of the object store. Would this method help? `getNSNameFromAdditionalState()`?

**Comment by thotz** [comment]
> The namespace of the cephcluster is the same as the namespace of the object store. Would this method help? `getNSNameFromAdditionalState()`?

That function was erroring out while handling OBC which were not provisioned by Rook. So I thought handling this check before fetching details about object store in the OBC

**Review by subhamkrai — COMMENTED** [review]
Just adding a comment that lots of CI are failing.

**Review by travisn — APPROVED** [review]

**Pull request merged**